### PR TITLE
Various performance improvements

### PR DESCRIPTION
**Pull request by BlobCodes**
This PR includes a few performance improvements that are very trivial and don't really need their own PR.

Performance improvements and other notes can be found in the commit messages.

The only change that isn't completely trivial is 99f8cd6326a98f40c49b32e39cd8a0cb3359508e, but that's really just Char#each_byte glued together with String.new.

### CODE REVIEW

**Comment by wontruefree** [comment]
since these are performance improvements do you have any benchmarks?

**Comment by BlobCodes** [comment]
> since these are performance improvements do you have any benchmarks?

I had benchmarks (the results are in the git commit message), but I don't have them anymore.

**Comment by j8r** [comment]
Yes, stating x% spees improvement without the actual code is quite meaningless.

**Comment by j8r** [comment]
Yes, stating x% speed improvement without the actual code is quite meaningless.

**Comment by BlobCodes** [comment]
[benchmark.zip](https://github.com/crystal-lang/crystal/files/7715288/benchmark.zip)
Benchmarks.
These benchmarks only test against one input (not cherrypicked), but it's probably better to test against different inputs.

**Comment by wontruefree** [comment]
here is an initial run for people on this thread. Improvements range from 1.03 times improvements to 2.61 times improvement.  I have run these a few times and Blowfish has a 0.01 variance and Char#to_s has a 0.3 variance.

```
$ ./script 
Benchmark 1: Blowfish (02a94043c2d7eb3767dbbfd5e9813d29f521a3ca)
old  18.48  ( 54.12ms) (± 4.12%)  5.13kB/op   1.03× slower
new  19.01  ( 52.61ms) (± 3.91%)  5.02kB/op        fastest
Benchmark 2: URI#request_target (9095f9bc670b693bc3eabb665ddfd31eefbdfea1)
old  11.58M ( 86.36ns) (± 6.03%)  224B/op   1.25× slower
new  14.46M ( 69.16ns) (± 6.93%)  144B/op        fastest
Benchmark 3: URI#resolve_path (e91ac9742385260fbdbc02b61ba24720792be775)
old   1.85M (540.38ns) (± 5.34%)  624B/op   1.22× slower
new   2.26M (442.15ns) (± 4.60%)  448B/op        fastest
Benchmark 4: URI#relativize_path (341d74943ab0b0f363284a395bb385f87e92d0a5)
old   1.22M (817.47ns) (± 6.51%)  916B/op   1.14× slower
new   1.40M (716.56ns) (± 5.10%)  805B/op        fastest
Benchmark 5: Char#to_s (99f8cd6326a98f40c49b32e39cd8a0cb3359508e)
old  37.87M ( 26.41ns) (± 5.86%)  48.0B/op   2.61× slower
new  98.92M ( 10.11ns) (± 6.88%)  16.0B/op        fastest
Benchmark 6: HTTP::Cookie#to_cookie_header (0877acf7c72c7c17a3818ae37a9ebfeb6f7037dc)
old  11.59M ( 86.26ns) (± 6.37%)  208B/op   1.24× slower
new  14.42M ( 69.37ns) (± 7.18%)  128B/op        fastest```

File: src/uri.cr [hunk 1]
```diff
@@ -221,7 +221,11 @@ class URI
   # uri.request_target # => "/?foo=bar"
   # ```
   def request_target : String
-    String.build do |str|
+    string_size = 1
+    string_size += @query ? @query.not_nil!.bytesize : -1
```
Comment (konovod) [inline src/uri.cr:225]: I think better way to deal with nillable properties is something like
```if q = @query 
  string_size += q.bytesize 
else
  string_size -= 1
end
```
On the other hand, procedure isn't thread-safe anyway.

**Comment by straight-shoota** [comment]
I'm pretty sure we want individual PRs for individual changes. Grouping related changes in different locations is fine. For example the different optimizations passing an initial size to `String.build`.
But please don't put unrelated changes in unrelated locations into a single PR. Having more PRs doesn't cost anything, but it helps to keep the discussions separate and the commit history organized. Thanks.

---
It appears the majority of changes are related to `String.build`, so we could keep them all in this PR and revert all the others, creating separate PRs for them. But I think it's better to create new individual PRs and close this one.

**Comment by BlobCodes** [comment]
Cherry-picked into #11592, #11593 and #11594

**Pull request closed without merging**